### DISCUSSION TITLE
special vdevs need plp? (already have slog w plp)

### DISCUSSION THREAD

**Discussion by ghost** [openzfs/zfs#13444] [Q&A] [answered]
I have a slog with power loss protection and I'm wondering if special vdevs need plp too, when you already have a slog with plp? Does every write to special go through the slog first?

This also roused another question in me, which is; what happens to async writes in general on power loss, when you have a proper slog? I assume all writes in zfs actually become sync writes at some point? Or are async writes actually never used in practice except for caching?

**Comment by amotin** [comment, marked as answer]
ZFS respects application requests.  Sync writes are first written into SLOG and then as part of TXG commit written to the main pool (or special vdev, there is no difference).  Async writes are waiting for the TXG commit in memory, and if power is lost before the last synchronize cache request of the commit -- they are lost.

You should probably read what exactly the PLP promise you for specific vendor/model.  ZFS does not need every write to the device to be immediately persistent.  It sends synchronize cache requests when required, and once those requests are completed -- data must be safe on non-volatile storage.  SSDs with supercaps or other PLP techniques may be able to handle those requests much faster, like 10us instead of 500us, that is critical for SLOG, but not really for other vdevs if SLOG is present.

**Reply by ghost** [reply]
**"Sync writes are first written into SLOG and then as part of TXG commit written to the main pool (or special vdev, there is no difference). Async writes ..."**
(as someone who doesn't know zfs internals) This sounds like you might be only talking about the small files being optionally stored on the special vdevs, I'm unsure. 

**"that is critical for SLOG, but not really for other vdevs if SLOG is present."**
So this includes the metadata being stored on special vdevs? (to be 100% sure)

**Reply by bghira** [reply]
there's also the `logbias` property that modulates how the I/O is submitted:

```
     logbias=latency|throughput
       Provide a hint to ZFS about handling of synchronous requests in this dataset.  If logbias is set to latency (the default), ZFS will use pool log devices (if configured) to handle the requests at low latency.  If logbias is set to throughput,
       ZFS will not use configured pool log devices.  ZFS will instead optimize synchronous operations for global pool throughput and efficient use of resources.

```

**Reply by amotin** [reply]
> So this includes the metadata being stored on special vdevs? (to be 100% sure)

Special vdevs are not so special. They just have priority for storing some categories of data, like metadata or small blocks.  In all other aspects they are identical to normal vdevs.

@bghira is right in his comment that setting logbias=throughput may indeed put synchronous writes into the normal (and probably special) vdevs, that may increase their requirements to sync write latency, but it is pretty special mode for which I personally haven't found a good use case yet.

**Reply by ghost** [reply]
Thank you everyone for the amazing answers <3
